feat: Implement builtin mode for keepassxc

bakito opened the pull request:
Implements: #4148

twpayne commented on .gitignore:
```diff
@@ -15,3 +15,4 @@
 /dist
 /resource_windows_*.syso
 /versioninfo.json
+/.idea
```
Please remove this, user-specific configuration should go in your global `.gitignore` file.

twpayne commented on go.mod:
```diff
@@ -1,8 +1,6 @@
-go 1.22.7
-
-toolchain go1.23.3
+go 1.23.4
```
I noticed that github.com/tobischo/gokeepasslib/v3 requires Go 1.23. FYI I'm working on a separate PR to switch chezmoi over to Go 1.23. I'll ask you to rebase this PR when its done.

twpayne reviewed (commented):
Thanks for this!

Instead of a new set of template functions, I think this different ["mode" for opening KeepassXC databases](https://www.chezmoi.io/reference/templates/keepassxc-functions/). Currently chemzoi supports "cache-password" and "open" modes. I think this should be a new "builtin" mode.

What do you think?

bakito commented:
> Thanks for this!
> 
> Instead of a new set of template functions, I think this different ["mode" for opening KeepassXC databases](https://www.chezmoi.io/reference/templates/keepassxc-functions/). Currently chemzoi supports "cache-password" and "open" modes. I think this should be a new "builtin" mode.
> 
> What do you think?

Sounds reasonable.
I modified the implementation and added the functions for the new mode.

bakito commented on internal/cmd/keepassxctemplatefuncs.go:
```diff
@@ -100,6 +113,18 @@ func (c *Config) keepassxcTemplateFunc(entry string) map[string]string {
 func (c *Config) keepassxcAttributeTemplateFunc(entry, attribute string) string {
+	if c.Keepassxc.Mode == keepassxcModeBuiltin {
+		// builtin stores attributes in cache
+		if data, ok := c.Keepassxc.cache[entry]; ok {
```
As the library handles default values the same as attributes, both functions use the same cache in builtin mode.
I don't think that i makes sense to keep separate caches in this case.

bakito commented on internal/cmd/keepassxctemplatefuncs.go:
```diff
@@ -329,3 +354,82 @@ func (c *Config) keepassxcClose() error {
 }
+
+// keepassxcBuiltinExtractValues extract builtin values
+func (c *Config) keepassxcBuiltinExtractValues(
```
This implementation reads all values from the keepass database to reduce access to the file itself.
If this might blow the memory I can change the implementation to only cache the requested values.

halostatue replied:
We don't know how large any given keepass database will be, so the memory is potentially uncontrollable this way. More problematic, though, is that this decrypts the entire database and puts it in memory. chezmoi doesn't do anything specific to harden its memory accesses (I’m not sure what one could do with Go), so this represents a much larger attack surface on a running instance of chezmoi.

If it isn't too much additional work, I think that only entries requested should be cached because to do otherwise would change chezmoi's security profile.

bakito replied:
@halostatue @twpayne I've modified the code to only cache required entries

twpayne reviewed (commented):
Thank you, this is looking excellent, and I will do a full review after the weekend.

Please also check the [contributing guide](https://www.chezmoi.io/developer-guide/contributing-changes/) for low-hanging fruit.

But, thank you, this PR is excellent.

bakito commented on internal/cmd/keepassxctemplatefuncs_test.go:
```diff
@@ -117,6 +117,7 @@ func TestKeepassxcTemplateFuncs(t *testing.T) {
 	for _, mode := range []keepassxcMode{
 		keepassxcModeCachePassword,
 		keepassxcModeOpen,
+		keepassxcModeBuiltin,
```
I'm having some issues on running this test.
When only `keepassxcModeBuiltin` mode is enabled, the test is running. 
But when the other 2 are also enabled (or only the 2 other) the test fails.

I can't figure out the issue here. Might be some precondition in my environment.

```
Enter password to encrypt database (optional): 
Repeat password: 
Successfully created new database.
Enter password to unlock /tmp/TestKeepassxcTemplateFuncs1798468072/001/KeePassXC Passwords.kdbx: 
Successfully added group test group.
Enter password to unlock /tmp/TestKeepassxcTemplateFuncs1798468072/001/KeePassXC Passwords.kdbx: 
Enter password for new entry: 
Successfully added entry test entry.
Enter password to unlock /tmp/TestKeepassxcTemplateFuncs1798468072/001/KeePassXC Passwords.kdbx: 
Successfully imported attachment /tmp/TestKeepassxcTemplateFuncs1798468072/001/import file as test / attachment name to entry test group/test entry.
error: Permission denied
--- FAIL: TestKeepassxcTemplateFuncs (0.37s)
    --- FAIL: TestKeepassxcTemplateFuncs/cache-password (0.01s)
        --- FAIL: TestKeepassxcTemplateFuncs/cache-password/correct_password (0.01s)
panic: /home/bakito/bin/keepassxc-cli show '/tmp/TestKeepassxcTemplateFuncs1798468072/001/KeePassXC Passwords.kdbx' --quiet --show-protected 'test group/test entry': exit status 1 [recovered]
	panic: /home/bakito/bin/keepassxc-cli show '/tmp/TestKeepassxcTemplateFuncs1798468072/001/KeePassXC Passwords.kdbx' --quiet --show-protected 'test group/test entry': exit status 1

goroutine 415 [running]:
testing.tRunner.func1.2({0x17a2560, 0xc006f6e370})
	/home/bakito/.gimme/versions/go1.23.4.linux.amd64/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	/home/bakito/.gimme/versions/go1.23.4.linux.amd64/src/testing/testing.go:1635 +0x35e
panic({0x17a2560?, 0xc006f6e370?})
	/home/bakito/.gimme/versions/go1.23.4.linux.amd64/src/runtime/panic.go:785 +0x132
github.com/twpayne/chezmoi/v2/internal/cmd.mustValue[...](...)
	/home/bakito/go/src/github.com/bakito/chezmoi/internal/cmd/cmd.go:254
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).keepassxcTemplateFunc(0xc0073aa008, {0xc008243e18, 0x15})
	/home/bakito/go/src/github.com/bakito/chezmoi/internal/cmd/keepassxctemplatefuncs.go:103 +0x24a
github.com/twpayne/chezmoi/v2/internal/cmd.TestKeepassxcTemplateFuncs.func1.1(0xc0072c3380)
	/home/bakito/go/src/github.com/bakito/chezmoi/internal/cmd/keepassxctemplatefuncs_test.go:130 +0x1fa
testing.tRunner(0xc0072c3380, 0xc00737b170)
	/home/bakito/.gimme/versions/go1.23.4.linux.amd64/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 414
	/home/bakito/.gimme/versions/go1.23.4.linux.amd64/src/testing/testing.go:1743 +0x390
```

twpayne replied:
I think you've fixed this, right? I don't see this when I run the tests now.

bakito replied:
It seems only to be ann issue on my machine. As the pre existing tests do not run.
If the, run all on your side, thats fine for me.

halostatue commented on assets/chezmoi.io/docs/reference/templates/keepassxc-functions/index.md:
```diff
@@ -18,3 +18,6 @@ cache it for the duration of chezmoi's execution. Setting `keepassxc.mode` to
 values from KeePassXC.
+When setting the `keepassxc.mode` to `builtin`, chezmoi uses a builtin library to
+access a keepassxc database. This mode becomes handy if the keepassxc-cli is not
+available.
```
```markdown

When setting `keepassxc.mode` to `builtin`, chezmoi uses a builtin library to
access a keepassxc database, which can be handy if `keepassxc-cli` is not
available. Some features available in KeePassXC such as Yubikey-enhanced
encryption, may not be available with builtin support.
```

Note the extra blank line at the top.

bakito replied:
> ```
> When setting `keepassxc.mode` to `builtin`, chezmoi uses a builtin library to
> access a keepassxc database, which can be handy if `keepassxc-cli` is not
> available. Some features available in KeePassXC such as Yubikey-enhanced
> encryption, may not be available with builtin support.
> ```
> 
> Note the extra blank line at the top.

Not sure what you mean wit that. What should be changed?

halostatue replied:
In your original text, you had two paragraphs right next to each other. They should be separated by a blank line.

With the rest of it, I am suggesting a change.

twpayne commented:
Thank you, this is looking excellent. Once you've made the small changes to the documentation, this should be ready to merge.

halostatue commented on assets/chezmoi.io/docs/reference/templates/keepassxc-functions/index.md:
```diff
@@ -18,3 +18,8 @@ cache it for the duration of chezmoi's execution. Setting `keepassxc.mode` to
+
+When setting `keepassxc.mode` to `builtin`, chezmoi uses a builtin library to
+access a keepassxc database, which can be handy if `keepassxc-cli` is not
+available. Some features available in KeePassXC such as Yubikey-enhanced
+encryption, may not be available with builtin support.
```
The last sentence needs a slight tweak:

```suggestion
available. Some KeePassXC features (such as Yubikey-enhanced
encryption) may not be available with builtin support.
```

halostatue commented on internal/cmd/keepassxctemplatefuncs.go:
```diff
@@ -76,30 +78,57 @@ func (c *Config) keepassxcAttachmentTemplateFunc(entry, name string) string {
+		if data, ok := c.Keepassxc.cache[entry]; ok {
+			return data[name]
+		}
+		panic(fmt.Sprintf("attachment %s of entry %s not found", entry, name))
```
The `%s` parameters are reversed. The message isn't *quite* right, because this is only checking that `entry` is present, not that `entry[name]` is present.

```suggestion
		panic(fmt.Sprintf("entry %s attachment %s not found", entry, name))
```

halostatue commented on internal/cmd/keepassxctemplatefuncs.go:
```diff
@@ -76,30 +78,57 @@ func (c *Config) keepassxcAttachmentTemplateFunc(entry, name string) string {
+	case keepassxcModeBuiltin:
+		c.Keepassxc.cache[entry] = c.keepassxcBuiltinExtractValues(entry, keepassxcBuiltinMapAttachmentCache)
+		if data, ok := c.Keepassxc.cache[entry]; ok {
+			return data[name]
```
Shouldn't we look to see if `name` exists in `data` before trying to return the value? The value would be blank otherwise, which feels like it is incorrect.

halostatue reviewed (changes requested):
Looks good overall. Just a couple of comments and only one "required" change (for the docs).

twpayne commented:
@bakito Thanks for making the changes. Please can you squash everything into a single commit, and then this should be good to go.

twpayne commented:
Thank you @bakito for your excellent work and patience here!

bakito commented:
@twpayne I introduced a lint issue in the last change. Sorry for that.
#4166 provides a fix
